File: src/main/scala/chapter02/MnMcount.scala
package chapter02

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

/**
 * Usage: MnMcount <mnm_file_dataset>
 */
object MnMcount {
  def main(args: Array[String]) {
    val spark = SparkSession
      .builder
      .appName("MnMCount")
      .getOrCreate()

    if (args.length < 1) {
      print("Usage: MnMcount <mnm_file_dataset>")
      sys.exit(1)
    }

    // Get the M&M data set filename
    val mnmFile = args(0)

    // Read the file into a Spark DataFrame
    val mnmDF = spark.read.format("csv")
      .option("header", "true")
      .option("inferSchema", "true")
      .load(mnmFile)

    // Aggregate counts of all colors and groupBy() State and Color
    // orderBy() in descending order
    val countMnMDF = mnmDF
      .select("State", "Color", "Count")
      .groupBy("State", "Color")
      .agg(count("Count").alias("Total"))
      .orderBy(desc("Total"))

    countMnMDF.show(60)
    println(s"Total Rows = ${countMnMDF.count()}")
    println()

    // Aggregate counts for California only
    val caCountMnMDF = mnmDF
      .select("State", "Color", "Count")
      .where(col("State") === "CA")
      .groupBy("State", "Color")
      .agg(count("Count").alias("Total"))
      .orderBy(desc("Total"))

    caCountMnMDF.show(10)

    spark.stop()
  }
}

File: python/chapter02/mnmcount.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import *

import sys

if len(sys.argv) != 2:
    print("Usage: mnmcount <mnm_dataset_file>")
    sys.exit(1)

mnm_file = sys.argv[1]

spark = SparkSession.builder.appName("MnMCount").getOrCreate()

# Leer el CSV
mnm_df = (spark.read.format("csv")
          .option("header", "true")
          .option("inferSchema", "true")
          .load(mnm_file))

# Agregar por estado y color
count_mnm_df = (mnm_df
                .groupBy("State", "Color")
                .agg(count("Count").alias("Total"))
                .orderBy(desc("Total")))

count_mnm_df.show(60)
print(f"Total Rows = {count_mnm_df.count()}")

# Filtrar California
ca_count = (mnm_df
            .where(col("State") == "CA")
            .groupBy("State", "Color")
            .agg(count("Count").alias("Total"))
            .orderBy(desc("Total")))

ca_count.show(10)

spark.stop()
